README
true version 1.0.1
==============================

automatically return a true value when a file is required

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

    B::Hooks::OP::Annotation
    B::Hooks::OP::Check
    Devel::StackTrace
    ExtUtils::Depends

COPYRIGHT AND LICENCE

Copyright (c) 2010-2020 by chocolateboy.

This is free software; you can redistribute it and/or modify it under the
terms of the Artistic License 2.0.


Changes
Revision history for Perl extension true.

1.0.2 TBD

    - code cleanup
    - documentation tweaks
    - update ppport.h from 3.44 -> 3.62

1.0.1 2019-03-11

    - add missing test for RT-124745 to the manifest

1.0.0 2019-03-11

    - RT-124745: fix breakage when using Moo/Moose and Function::Parameters on perl < 5.24 (thanks, Smylers)
      - add t/rt-124745.t
    - update ppport.h from 3.19 -> 3.44

0.18 2011-04-18

    - fix EUMM workaround (Christian Walde)

0.17 2011-04-18

    - RT-65586: work around EUMM issue on ActivePerl (Christian Walde and Schwern)
    - Makefile.PL fix

0.16 2011-02-14

    - true::VERSION: make the version static

0.15 2011-02-14

    - make true::VERSION a real module so dependency checkers can find it (thanks, Schwern)

0.14 2011-02-11

    - create $true::VERSION::VERSION to work around a true dependency issue (thanks, Schwern)

0.13 2011-01-27

    - true.pm: more robust detection of the currently-compiling file
        - should fix Windows failures - thanks, Alexandr Ciornii and Schwern
    - eval.t: confirm string evals aren't affected
    - clean up after compile-time exceptions: preliminary tests
    - add repository link

0.12 2010-08-17

    - ExtUtils::Depends fix
    - doc tweak

0.11 2010-08-17

    - add missing dependency
    - add support/documentation for unimport
      - add unimport.t
    - add leak.t
    - doc tweakage

0.10 2010-08-17

    - rm caveats
    - more tests

0.03 2010-01-25

    - doc tweak

0.02 2010-01-25

    - POD fmt fix
    - add pod.t

0.01 2010-01-25

    - initial release


t/lib/Good.pm
package Good;

use Contemporary::Perl;

sub Good { 'Good' };


t/lib/Contemporary/Perl.pm
package Contemporary::Perl;

use strict;
use warnings;
use true ();

sub import {
    strict->import();
    warnings->import();
    true->import();
}

1;


t/lib/Contemporary/Perl/Subclass.pm
package Contemporary::Perl::Subclass;

use strict;
use warnings;
use base qw(Contemporary::Perl);
use true; # after use base qw(...)

sub import {
    my $class = shift;
    $class->SUPER::import(@_);
}


lib/true.pm
package true;

use strict;
use warnings;

use B::Hooks::OP::Annotation;
use B::Hooks::OP::Check;
use Devel::StackTrace;
use XSLoader;

# XXX this declaration must be on a single line
# https://metacpan.org/pod/version#How-to-declare()-a-dotted-decimal-version
use version 0.77; our $VERSION = version->declare('v1.0.1');

our %TRUE;

# set the logger from the first rule which matches the $TRUE_PM_DEBUG
# environment variable:
#
#  - warn with a backtrace (cluck) if it's an integer > 1
#  - warn (without a backtrace) if it's truthy
#  - otherwise (default), do nothing
#
our $DEBUG = do {
    my $debug = $ENV{TRUE_PM_DEBUG} || '0';
    $debug = $debug =~ /^\d+$/ ? int($debug) : 1;
    $debug && ($debug > 1 ? do { require Carp; \&Carp::cluck } : sub { warn(@_) });
};

XSLoader::load(__PACKAGE__, $VERSION);

sub _debug($) {
    $DEBUG->(@_, $/) if ($DEBUG);
}

# return the full path of the file that's currently being compiled.
# XXX CopFILE(&PL_compiling) gives the wrong result here (it works in the
# OP-checker in the XS).
sub ccfile() {
    my ($file, $source, $line);
    my $trace = Devel::StackTrace->new;

    # find the innermost require frame
    #
    # for "use Foo::Bar" or "require Foo::Bar", the evaltext contains
    # "Foo/Bar.pm", and the filename/line refer to the file where the
    # use/require statement appeared.

    # work from the innermost frame out
    while (my $frame = $trace->next_frame) {
        next unless ($frame->is_require);

        my $required = $frame->evaltext;

        if (defined($file = $INC{$required})) {
            $source = $frame->filename;
            $line = $frame->line;
        } else { # shouldn't happen
            warn "true: can't find required file ($required) in \%INC";
        }

        last;
    }

    return wantarray ? ($file, $source, $line) : $file;
}

sub import {
    my ($file, $source, $line) = ccfile();

    if (defined($file) && not($TRUE{$file})) {
        $TRUE{$file} = 1;
        _debug "true: enabling true for $file at $source line $line";
        xs_enter();
    }
}

sub unimport {
    my ($file, $source, $line) = ccfile();

    if (defined($file) && $TRUE{$file}) {
        _debug "true: disabling true for $file at $source line $line";
        delete $TRUE{$file};
        xs_leave() unless (%TRUE);
    }
}

1;

__END__

=head1 NAME

true - automatically return a true value when a file is required

=head1 SYNOPSIS

  package Contemporary::Perl;

  use strict;
  use warnings;
  use true;

  sub import {
      strict->import();
      warnings->import();
      true->import();
  }

=head1 DESCRIPTION

Perl's C<require> builtin (and its C<use> wrapper) requires the files it loads
to return a true value. This is usually accomplished by placing a single

    1;

statement at the end of included scripts or modules. It's not onerous to add
but it's a speed bump on the Perl novice's road to enlightenment. In addition,
it appears to be a I<non-sequitur> to the uninitiated, leading some to attempt
to mitigate its appearance with a comment:

    1; # keep require happy

or:

    1; # Do not remove this line

or even:

    1; # Must end with this, because Perl is bogus.

This module packages this "return true" behaviour so that it doesn't need to be
written explicitly. It can be used directly, but it is intended to be invoked
from the C<import> method of a L<Modern::Perl|Modern::Perl>-style module that
enables modern Perl features and conveniences and cleans up legacy Perl warts.

=head2 METHODS

C<true> is file-scoped rather than lexically-scoped. Importing it anywhere in a
file (e.g. at the top-level or in a nested scope) causes that file to return true,
and unimporting it anywhere in a file restores the default behaviour. Redundant
imports/unimports are ignored.

=head3 import

Enable the "automatically return true" behaviour for the currently-compiling
file. This should typically be invoked from the C<import> method of a module
that loads C<true>. Code that uses this module solely on behalf of its callers
can load C<true> without importing it e.g.

    use true (); # don't import

    sub import {
        true->import();
    }

    1;

But there's nothing stopping a wrapper module also importing C<true> to obviate
its own need to explicitly return a true value:

    use true; # both load and import it

    sub import {
        true->import();
    }

    # no need to return true

=head3 unimport

Disable the "automatically return true" behaviour for the currently-compiling file.

=head2 EXPORTS

None by default.

=head1 NOTES

Because the unquoted name C<true> represents the boolean value C<true> in YAML,
the module name must be quoted when written as a dependency in META.yml. In cases
where this can't easily be done, a dependency can be declared on the package
L<true::VERSION>, which has the same version as C<true.pm>.

=head1 VERSION

1.0.1

=head1 SEE ALSO

=over

=item * L<latest>

=item * L<Modern::Perl>

=item * L<nonsense>

=item * L<perl5i>

=item * L<Toolkit>

=back

=head1 AUTHOR

chocolateboy <chocolate@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2010-2019 by chocolateboy.

This library is free software; you can redistribute it and/or modify it under the
terms of the L<Artistic License 2.0|https://www.opensource.org/licenses/artistic-license-2.0.php>.

=cut


Makefile.PL
use 5.008001;

use strict;
use warnings;

use Config;
use ExtUtils::Depends;
use ExtUtils::MakeMaker;
use File::Copy 'copy';

our $EUMM_VERSION = eval($ExtUtils::MakeMaker::VERSION);
our $OPTIMIZE;

if ($Config{gccversion}) {
    $OPTIMIZE = '-O3 -Wall -W';
    $OPTIMIZE .= ' -g -Wextra -Wdeclaration-after-statement' if (-d 'dev');
} elsif ($Config{osname} eq 'MSWin32') {
    $OPTIMIZE = '-O2 -W4';
} else {
    $OPTIMIZE = $Config{optimize};
}

our %XS_PREREQUISITES = (
    'B::Hooks::OP::Annotation' => '0.44',
    'B::Hooks::OP::Check'      => '0.22',
);

our %XS_DEPENDENCIES = ExtUtils::Depends->new(
    'true',
    keys(%XS_PREREQUISITES)
)->get_makefile_vars();

# test dependencies
our %BUILD_REQUIRES;

# Function::Parameters requires perl >= v5.14.0
if ($] >= 5.014000) {
    %BUILD_REQUIRES = (
        # https://rt.cpan.org/Ticket/Display.html?id=124745
        Moo => '2.003004', 'Function::Parameters' => '2.001003',
    );
}

sub ensure_activeperl_dep_files_exist($) {
    return if $^O ne 'MSWin32';
    return if $Config{make} !~ /dmake/;
    return if $Config{cc} !~ /gcc/;
    return if $Config{cf_email} !~ /ActiveState/;

    my $libs = shift;
    my @libs = split ' ', $libs;
    my @dirs  = grep { /^-L.+/ } @libs;
    my @files = grep { /^-l.+/ } @libs;

    s/^-L// for @dirs;
    s/^-l// for @files;

    for my $dir (@dirs) {
        for my $file (@files) {
            my $base = "$dir/$file";
            my $want = "$base.a";
            my $maybe = "$base.lib";

            # everything OK: next
            next if -f $want;

            # if neither file exists, either this one is elsewhere
            # or we have a problem: next
            next unless -f $maybe;

            # .a is missing, but .lib exists, so just copy it over
            copy($maybe, $want);
        }
    }
}

ensure_activeperl_dep_files_exist($XS_DEPENDENCIES{LIBS});

my %META = (
    NAME          => 'true',
    VERSION_FROM  => 'lib/true.pm',
    PREREQ_PM => {
        %XS_PREREQUISITES,
        'Devel::StackTrace' => '2.03',
    },
    ABSTRACT_FROM => 'lib/true.pm',
    AUTHOR        => 'chocolateboy <chocolate@cpan.org>',
    LIBS          => [''],
    DEFINE        => '',
    INC           => '-I.',
    OPTIMIZE      => $OPTIMIZE,
    %XS_DEPENDENCIES,
);

if ($EUMM_VERSION >= 6.55_03) {
    $META{BUILD_REQUIRES} = \%BUILD_REQUIRES;
} else {
    $META{PREREQ_PM} = { %{ $META{PREREQ_PM} || {} }, %BUILD_REQUIRES };
}

if ($EUMM_VERSION >= 6.31) {
    $META{LICENSE} = 'artistic_2';
}

if ($EUMM_VERSION >= 6.46) {
    $META{META_MERGE} = {
        configure_requires => {
            'ExtUtils::Depends' => '0.405',
            %XS_PREREQUISITES
        },
        dynamic_config => 1, # [1]
        resources => {
            repository => 'https://github.com/chocolateboy/true',
            bugtracker => 'https://github.com/chocolateboy/true/issues',
        },
    };
}

# [1] make sure this script is always run (to correctly handle the
# conditional test dependency on Function::Parameters). this avoids signaling to
# clients that the static metadata in META.yml or META.json is sufficient for
# every build
#
# http://blogs.perl.org/users/neilb/2017/04/an-introduction-to-distribution-metadata.html
# https://github.com/Perl-Toolchain-Gang/CPAN-Meta/issues/112

if ($EUMM_VERSION >= 6.48) {
    $META{MIN_PERL_VERSION} = '5.008';
}

WriteMakefile(%META);
